vmm/src/lib.rs

```diff
@@ -1511,6 +1532,117 @@ impl Vmm {
             .map_err(|e| VmmActionError::NetworkConfig(ErrorKind::User, e))
     }
 
+    fn update_net_device(
+        &mut self,
+        new_cfg: NetworkInterfaceUpdateConfig,
+    ) -> std::result::Result<VmmData, VmmActionError> {
+        if !self.is_instance_initialized() {
+            // VM not started yet, so we only need to update the device configs, not the actual
+            // live device.
+            let old_cfg = self
+                .network_interface_configs
+                .iter_mut()
+                .find(|&&mut ref c| c.iface_id == new_cfg.iface_id)
+                .ok_or(VmmActionError::NetworkConfig(
+                    ErrorKind::User,
+                    NetworkInterfaceError::DeviceIdNotFound,
+                ))?;
+
+            // Check if we need to update the RX rate limiter.
+            if let Some(new_rlim_cfg) = new_cfg.rx_rate_limiter {
+                if let Some(ref mut old_rlim) = old_cfg.rx_rate_limiter {
+                    // We already have an RX rate limiter set, so we'll update it.
+                    old_rlim.update_buckets(
+                        new_rlim_cfg.bandwidth.map(|b| b.into_token_bucket()),
+                        new_rlim_cfg.ops.map(|b| b.into_token_bucket()),
+                    );
+                } else {
+                    // No old RX rate limiter; create one now.
+                    old_cfg.rx_rate_limiter =
+                        Some(new_rlim_cfg.into_rate_limiter().map_err(|e| {
+                            VmmActionError::NetworkConfig(
+                                ErrorKind::Internal,
+                                NetworkInterfaceError::RateLimiterError(e),
+                            )
+                        })?);
+                }
+            }
+
+            // Check if we need to update the TX rate limiter.
+            if let Some(new_rlim_cfg) = new_cfg.tx_rate_limiter {
```

Comment:
nit: this logic looks pretty much the same as the one above, can you move it into a helper function?

Reply:
Simplified the code a bit there.